### ISSUE TITLE
some bugs of examples\gradient-based-offpolicy\q_learning_offpolicy.py

### ISSUE THREAD

**Issue by Alisa742** [FilippoAiraldi/mpc-reinforcement-learning#10]
![Image](https://github.com/user-attachments/assets/c035df6a-abcf-4f9d-8b8f-ad8342603f21)

Hello! I am new to reinforcement learning, and try to understand how q_learning policy works but I run examples\gradient-based-offpolicy\q_learning_offpolicy.py, the outputs just like the above picture. I don't understand why there are valid rewards in validation step but get NAN in training step?
Moreover, I am curious to know if this package can be the foundation to design my vehicle stability control algorithm which gets environment data from CARSIM and be trained online?

**Comment by FilippoAiraldi** [comment]
If you pay attention to the source code, in particular here

https://github.com/FilippoAiraldi/mpc-reinforcement-learning/blob/070ba31839f02bee4e0305b318c5d92617e27935/src/mpcrl/agents/common/learning_agent.py#L260-L262

you will notice that the nans are intentional. Since the example is training the MPC in an off-policy fashion, the episodes' rewards are not influenced by the current MPC policy, as the training data is pulled from another (exploratory) policy. For this reason, I decided to let the logger print nans. On the other hand, during validation, we run the actual current MPC policy, and that's why we get valid rewards.

I am not knowledgable on CARSIM, so I cannot help you there. What I can tell you is that, as long as you can wrap your simulation environment in a Gymnasium Environment class, my library will be able to interact with it and, if everything goes smoothly, learn from it. The code does not support other environment interfaces for now.

**Comment by Alisa742** [comment]
thank you very much!



---Original---
From: "Filippo ***@***.***&gt;
Date: Sat, Jan 25, 2025 18:11 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [FilippoAiraldi/mpc-reinforcement-learning] some bugs ofexamples\gradient-based-offpolicy\q_learning_offpolicy.py (Issue #10)




 
Closed #10 as completed.
